File: go/internal/vectest/stream.go
package vectest

import (
	"encoding/json"
	"fmt"
	"testing"

	"github.com/elkpi/oxa/go/ir"
)

// StreamConverter is the face-implementation surface that RunStream drives.
// Native stream JSON is intentionally decoded and encoded by face-local test
// adapters, so this generic package never constructs provider wire types.
type StreamConverter interface {
	// Face is the vectors/ directory name of the face.
	Face() string
	DecodeNativeEvent(json.RawMessage) ([]ir.Event, error)
	FlushDecoder() ([]ir.Event, error)
	DecoderLosses() []ir.Loss
	ApplyIREvent(ir.Event) ([]json.RawMessage, []ir.Loss, error)
}

// streamVectorResetter is an optional test-adapter capability. RunStream calls
// it before each vector so that a converter remains isolated even when a prior
// vector ends with a conversion error.
type streamVectorResetter interface {
	ResetStreamVector()
}

// RunStream executes every stream golden vector for conv's face. It skips when
// the vectors repository cannot be located (dependency mode), but requires at
// least one stream vector when running from the monorepo.
func RunStream(t *testing.T, conv StreamConverter) {
	t.Helper()
	root := FindRepoRoot(".")
	if root == "" {
		t.Skip("repo root not found; vector tests skipped (dependency mode)")
	}
	vectors, err := LoadVectors(root, conv.Face(), "stream")
	if err != nil {
		t.Fatalf("load stream vectors for face %s: %v", conv.Face(), err)
	}
	if len(vectors) == 0 {
		t.Fatalf("no stream vectors found for face %s; the harness must execute at least one", conv.Face())
	}
	for _, v := range vectors {
		v := v
		t.Run(v.Name, func(t *testing.T) {
			if resetter, ok := conv.(streamVectorResetter); ok {
				resetter.ResetStreamVector()
			}
			switch v.Conversion {
			case "to-ir":
				actual, losses, err := runStreamToIR(conv, v.Input)
				if err != nil {
					t.Fatalf("decode stream failed: %v", err)
				}
				if err := CompareJSON(v.ExpectedIR, actual); err != nil {
					t.Errorf("expected_ir mismatch: %v\nexpected: %s\nactual:   %s", err, v.ExpectedIR, actual)
				}
				compareLosses(t, v, losses)
			case "from-ir":
				if _, err := unmarshalNativeEvents(v.ExpectedOut); err != nil {
					t.Fatalf("expected_output events envelope: %v", err)
				}
				actual, losses, err := runStreamFromIR(conv, v.Input)
				if err != nil {
					t.Fatalf("encode stream failed: %v", err)
				}
				if err := CompareJSON(v.ExpectedOut, actual); err != nil {
					t.Errorf("expected_output mismatch: %v\nexpected: %s\nactual:   %s", err, v.ExpectedOut, actual)
				}
				compareLosses(t, v, losses)
			default:
				t.Fatalf("unknown conversion %q", v.Conversion)
			}
		})
	}
	t.Logf("vectest: executed %d stream vectors for face %s", len(vectors), conv.Face())
}

func runStreamToIR(conv StreamConverter, input json.RawMessage) (json.RawMessage, []ir.Loss, error) {
	nativeEvents, err := unmarshalNativeEvents(input)
	if err != nil {
		return nil, nil, err
	}

	events := make([]ir.Event, 0, len(nativeEvents))
	for i, raw := range nativeEvents {
		decoded, err := conv.DecodeNativeEvent(raw)
		if err != nil {
			return nil, nil, fmt.Errorf("decode native event %d (%s): %w", i, boundedNativeEvent(raw), err)
		}
		events = append(events, decoded...)
	}
	flushed, err := conv.FlushDecoder()
	if err != nil {
		return nil, nil, fmt.Errorf("flush stream decoder: %w", err)
	}
	events = append(events, flushed...)
	losses := conv.DecoderLosses()

	out, err := ir.MarshalEventStream(&ir.EventStream{Events: events})
	if err != nil {
		return nil, nil, fmt.Errorf("marshal IR event stream: %w", err)
	}
	return out, losses, nil
}

func runStreamFromIR(conv StreamConverter, input json.RawMessage) (json.RawMessage, []ir.Loss, error) {
	stream, err := ir.UnmarshalEventStream(input)
	if err != nil {
		return nil, nil, fmt.Errorf("unmarshal IR event stream: %w", err)
	}

	nativeEvents := make([]json.RawMessage, 0, len(stream.Events))
	var losses []ir.Loss
	for i, ev := range stream.Events {
		encoded, reported, err := conv.ApplyIREvent(ev)
		if err != nil {
			return nil, nil, fmt.Errorf("apply IR event %d: %w", i, err)
		}
		nativeEvents = append(nativeEvents, encoded...)
		losses = append(losses, reported...)
	}
	out, err := json.Marshal(struct {
		Events []json.RawMessage `json:"events"`
	}{Events: nativeEvents})
	if err != nil {
		return nil, nil, fmt.Errorf("marshal native event envelope: %w", err)
	}
	return out, losses, nil
}

const maxNativeEventErrorBytes = 512

func boundedNativeEvent(raw json.RawMessage) string {
	if len(raw) <= maxNativeEventErrorBytes {
		return string(raw)
	}
	return string(raw[:maxNativeEventErrorBytes]) + "…"
}

func unmarshalNativeEvents(raw json.RawMessage) ([]json.RawMessage, error) {
	var envelope struct {
		Events *json.RawMessage `json:"events"`
	}
	if err := json.Unmarshal(raw, &envelope); err != nil {
		return nil, fmt.Errorf("unmarshal events envelope: %w", err)
	}
	if envelope.Events == nil {
		return nil, fmt.Errorf("missing events envelope")
	}
	var events []json.RawMessage
	if err := json.Unmarshal(*envelope.Events, &events); err != nil {
		return nil, fmt.Errorf("events is not an array: %w", err)
	}
	if events == nil {
		return nil, fmt.Errorf("events must be an array")
	}
	return events, nil
}

File: rust/crates/oxa-vectest/src/stream.rs
//! The stream vector runner: native stream JSON is decoded and encoded by
//! face-local test adapters, so this generic module never constructs
//! provider wire types.

use std::path::Path;

use oxa_ir::{Event, EventStream, Loss};
use serde_json::Value;

use crate::compare::compare_json;
use crate::load::{find_repo_root, load_vectors};
use crate::run::{Outcome, Report, compare_reported_losses, json_text};

/// The face-implementation surface that [`run_stream`] drives. Adapters are
/// stateful (they retain the face decoder/encoder), so methods take `&mut
/// self`.
pub trait StreamConverter {
    /// The vectors/ directory name of the face.
    fn face(&self) -> &'static str;
    /// Decodes one native event (one element of `input.events`).
    fn decode_native_event(&mut self, event: &Value) -> Result<Vec<Event>, String>;
    /// Supplies decoder terminal output.
    fn flush_decoder(&mut self) -> Result<Vec<Event>, String>;
    /// Returns the full loss list; the runner calls it only after all native
    /// events and [`StreamConverter::flush_decoder`] have run.
    fn decoder_losses(&self) -> Vec<Loss>;
    /// Applies one canonical IR event, returning ordered native event
    /// documents plus reported losses.
    fn apply_ir_event(&mut self, event: &Event) -> Result<(Vec<Value>, Vec<Loss>), String>;
    /// Optional per-vector state reset so a converter remains isolated even
    /// when a prior vector ends with a conversion error. Defaults to a no-op.
    fn reset_stream_vector(&mut self) {}
}

/// Executes every stream golden vector for the converter's face, locating
/// the repo root from the current working directory. Requires at least one
/// stream vector when running from the monorepo.
pub fn run_stream(conv: &mut dyn StreamConverter) -> Result<Outcome, String> {
    run_stream_in(Path::new("."), conv)
}

/// [`run_stream`] with an explicit starting directory for the repo-root walk.
pub fn run_stream_in(dir: &Path, conv: &mut dyn StreamConverter) -> Result<Outcome, String> {
    let Some(root) = find_repo_root(dir) else {
        return Ok(Outcome::Skipped);
    };
    let vectors = load_vectors(&root, conv.face(), "stream")?;
    if vectors.is_empty() {
        return Err(format!(
            "no stream vectors found for face {}; the harness must execute at least one",
            conv.face()
        ));
    }
    let mut report = Report::default();
    for vector in &vectors {
        conv.reset_stream_vector();
        match vector.conversion.as_str() {
            "to-ir" => {
                let Some(expected) = vector.expected_ir.as_ref() else {
                    report.failures.push((
                        vector.name.clone(),
                        "to-ir vector is missing expected_ir".to_string(),
                    ));
                    continue;
                };
                match stream_to_ir(conv, &vector.input) {
                    Ok((actual, losses)) => {
                        if let Err(err) = compare_json(expected, &actual) {
                            report.failures.push((
                                vector.name.clone(),
                                format!("expected_ir mismatch: {err}"),
                            ));
                        }
                        compare_reported_losses(vector, &losses, &mut report);
                    }
                    Err(err) => report
                        .failures
                        .push((vector.name.clone(), format!("decode stream failed: {err}"))),
                }
            }
            "from-ir" => {
                let Some(expected) = vector.expected_output.as_ref() else {
                    report.failures.push((
                        vector.name.clone(),
                        "from-ir vector is missing expected_output".to_string(),
                    ));
                    continue;
                };
                if let Err(err) = native_events(expected) {
                    report.failures.push((
                        vector.name.clone(),
                        format!("expected_output events envelope: {err}"),
                    ));
                    continue;
                }
                match stream_from_ir(conv, &vector.input) {
                    Ok((actual, losses)) => {
                        if let Err(err) = compare_json(expected, &actual) {
                            report.failures.push((
                                vector.name.clone(),
                                format!("expected_output mismatch: {err}"),
                            ));
                        }
                        compare_reported_losses(vector, &losses, &mut report);
                    }
                    Err(err) => report
                        .failures
                        .push((vector.name.clone(), format!("encode stream failed: {err}"))),
                }
            }
            other => report
                .failures
                .push((vector.name.clone(), format!("unknown conversion {other:?}"))),
        }
    }
    report.executed = vectors.len();
    Ok(Outcome::Ran(report))
}

/// Feeds every native event of `input` through the converter, appends the
/// flush output, and returns the canonical IR event stream document plus the
/// decoder losses.
pub fn stream_to_ir(
    conv: &mut dyn StreamConverter,
    input: &Value,
) -> Result<(Value, Vec<Loss>), String> {
    let native_events = native_events(input)?;
    let mut events = Vec::new();
    for (index, raw) in native_events.iter().enumerate() {
        let decoded = conv
            .decode_native_event(raw)
            .map_err(|err| format!("decode native event {index} ({}): {err}", bounded(raw)))?;
        events.extend(decoded);
    }
    let flushed = conv
        .flush_decoder()
        .map_err(|err| format!("flush stream decoder: {err}"))?;
    events.extend(flushed);
    let losses = conv.decoder_losses();
    let doc = oxa_ir::to_json(&EventStream { events }).map_err(|err| err.to_string())?;
    let doc: Value = serde_json::from_str(&doc).map_err(|err| err.to_string())?;
    Ok((doc, losses))
}

/// Applies every IR event of the `input` stream through the converter and
/// returns the native `{"events": [...]}` envelope plus the concatenated
/// encoder losses.
pub fn stream_from_ir(
    conv: &mut dyn StreamConverter,
    input: &Value,
) -> Result<(Value, Vec<Loss>), String> {
    let stream: EventStream = oxa_ir::from_json(&json_text(input))
        .map_err(|err| format!("unmarshal IR event stream: {err}"))?;
    let mut native_events = Vec::new();
    let mut losses = Vec::new();
    for (index, event) in stream.events.iter().enumerate() {
        let (encoded, reported) = conv
            .apply_ir_event(event)
            .map_err(|err| format!("apply IR event {index}: {err}"))?;
        native_events.extend(encoded);
        losses.extend(reported);
    }
    Ok((serde_json::json!({ "events": native_events }), losses))
}

const MAX_NATIVE_EVENT_ERROR_CHARS: usize = 512;

fn bounded(value: &Value) -> String {
    let text = json_text(value);
    if text.chars().count() <= MAX_NATIVE_EVENT_ERROR_CHARS {
        return text;
    }
    let mut bounded: String = text.chars().take(MAX_NATIVE_EVENT_ERROR_CHARS).collect();
    bounded.push('…');
    bounded
}

fn native_events(input: &Value) -> Result<Vec<Value>, String> {
    let Some(envelope) = input.as_object() else {
        return Err("input is not an object".to_string());
    };
    let Some(events) = envelope.get("events") else {
        return Err("missing events envelope".to_string());
    };
    let Some(events) = events.as_array() else {
        return Err("events is not an array".to_string());
    };
    Ok(events.clone())
}
